mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java

```diff
@@ -815,12 +818,17 @@ private void createUpdateEntityInfoIndexes(EntityInfo ei, Metadata md) {
                     newIndex.put(Translator.translatePath(field), p.isDesc() ? -1 : 1);
                 }
                 BasicDBObject options = new BasicDBObject("unique", index.isUnique());
-                // if index is unique also make it a sparse index, so we can have non-required unique fields
-                options.append("sparse", index.isUnique());
+                // if index is unique and non-partial, also make it a sparse index, so we can have non-required unique fields
+                options.append("sparse", index.isUnique() && !index.getProperties().containsKey(PARTIAL_FILTER_EXPRESSION_OPTION_NAME));
                 if (index.getName() != null && index.getName().trim().length() > 0) {
                     options.append("name", index.getName().trim());
                 }
                 options.append("background", true);
+                // partial index
+                if (index.getProperties().containsKey(PARTIAL_FILTER_EXPRESSION_OPTION_NAME)) {
+                    DBObject filter = (DBObject) JSON.parse(index.getProperties().get(PARTIAL_FILTER_EXPRESSION_OPTION_NAME).toString());
```

Comment:
iirc, this is not how the metadataparser will parse this. If the property is:

{ field: { .. } }

then, the property will return a Map, with a key "field", and value of another map, and so on.

You need to recursively process this and build a BasicDBObject from that, instead of using json parser. That could be useful, so you can add that as a static method to PropertyParser in core.